grpclb/src/main/java/io/grpc/grpclb/CachedSubchannelPool.java

```diff
@@ -23,65 +23,77 @@
 import io.grpc.ConnectivityStateInfo;
 import io.grpc.EquivalentAddressGroup;
-import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer.CreateSubchannelArgs;
 import io.grpc.LoadBalancer.Helper;
 import io.grpc.LoadBalancer.Subchannel;
+import io.grpc.LoadBalancer.SubchannelStateListener;
 import io.grpc.SynchronizationContext.ScheduledHandle;
 import java.util.HashMap;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * A {@link SubchannelPool} that keeps returned {@link Subchannel}s for a given time before it's
  * shut down by the pool.
  */
 final class CachedSubchannelPool implements SubchannelPool {
+  private static final Attributes.Key<AtomicReference<SubchannelStateListener>> STATE_LISTENER =
+      Attributes.Key.create("io.grpc.grpclb.CachedSubchannelPool.stateListener");
+
   private final HashMap<EquivalentAddressGroup, CacheEntry> cache =
       new HashMap<>();
 
   private Helper helper;
-  private LoadBalancer lb;
 
   @VisibleForTesting
   static final long SHUTDOWN_TIMEOUT_MS = 10000;
 
   @Override
-  public void init(Helper helper, LoadBalancer lb) {
+  public void init(Helper helper) {
     this.helper = checkNotNull(helper, "helper");
-    this.lb = checkNotNull(lb, "lb");
   }
 
   @Override
   public Subchannel takeOrCreateSubchannel(
-      EquivalentAddressGroup eag, Attributes defaultAttributes) {
+      EquivalentAddressGroup eag, Attributes defaultAttributes, SubchannelStateListener listener) {
     final CacheEntry entry = cache.remove(eag);
     final Subchannel subchannel;
     if (entry == null) {
-      subchannel = helper.createSubchannel(eag, defaultAttributes);
+      final Attributes attrs = defaultAttributes.toBuilder()
+          .set(STATE_LISTENER, new AtomicReference<>(listener)).build();
+      subchannel = helper.createSubchannel(CreateSubchannelArgs.newBuilder()
+          .setAddresses(eag)
+          .setAttributes(attrs)
+          .setStateListener(new SubchannelStateListener() {
+              @Override
+              public void onSubchannelState(Subchannel subchannel, ConnectivityStateInfo newState) {
+                CacheEntry cached = cache.get(subchannel.getAddresses());
+                if (cached == null || cached.subchannel != subchannel) {
+                  // Given subchannel is not cached, notify the listener
+                  attrs.get(STATE_LISTENER).get().onSubchannelState(subchannel, newState);
+                } else {
+                  cached.state = newState;
+                }
+              }
+            })
+          .build());
     } else {
       subchannel = entry.subchannel;
       entry.shutdownTimer.cancel();
-      // Make the balancer up-to-date with the latest state in case it has changed while it's
-      // in the cache.
+      subchannel.getAttributes().get(STATE_LISTENER).set(listener);
+      // Make the listener up-to-date with the latest state in case it has changed while it's in the
+      // cache.
```

Comment:
Is the whole method in syncContext? Why not run `listener.onSubchannelState(subchannel, entry.state)` directly, so that there will be no state change?

Reply:
`onSubchannelState()` is not re-entrant with the other logic from `GrpclbState` when called by channel impl. Although reentrancy may not cause a problem, it's better to keep the baheavior closer to with channel impl.

That said, I think there is a problem with the current code. If channel impl schedules an update before the task on the next line while this method is running, the update from the channel, which is newer, would be overwritten with the older value by the latter task. I went with a different approach of the implementation that always track Subchannels and their states, whether in or out of the pool, which ends up simpler.